ultimatepython/classes/encapsulation.py

```diff
@@ -0,0 +1,108 @@
+"""
+Encapsulation is a feature of OOP that allows you to hide the
+implementation details of a class from its users.
+Encapsulation allows us to limit the access of certain attributes
+within a class. This prevents users from directly accessing and modifying such
+attributes from outside the class. Instead, users must use methods to access and
+modify attributes.
+"""
+import secrets
+
+
+class BankAccount:
+    def __init__(self, account_holder_name):
+        """
+        In python, a class attribute can be made private by prefixing it with two underscores.
+        This makes it inaccessible to users outside the class.
+        By default, class attributes are public. Therefore, they can be accessed and modified
+        outside of the class.
+
+        Here, account_number and balace are private while account_holder_name is public.
+        """
+        self.account_holder_name = account_holder_name
+        """
+        The account number is generated automaticallyusing the randint function from
+        the random module when a new instance of the class is created.
+        The balance is set to 0.0 by default.
+        """
+        self.__account_number = secrets.randbelow(10**10)  # generate a random account number of 10 digits.
+        self.__balance = 0.0
+
+    def deposit(self, balance):
+        self.__balance += balance  # add the deposited amount to the balance.
+
+    def withdraw(self, balance):
+        self.__balance -= balance
+
+    def get_balance(self):
+        return self.__balance
+
+    def get_account_number(self):
+        """
+        The account number is generated randomly when a new instance of the class is created.
+        Since the attribute is also private, it cannot be accessed directly from outside the class.
+        The get_account_number method allows you to access the account number outside of the class.
+        But since we do not define a setter method for this variable, we cannot modify it outside the class.
+        Therefore, the account number generated while creating an object of the BankAccount class cannot be changed
+        but can only be read using this function.
+        """
+        return self.__account_number
+
+    def __set_account_number(self, number):
+        """
+        This is a private method. Similar to private variables,
+        private methods also cannot be accessed outside the class.
+        """
+        self.__account_number = number
+
+    def delete_account(self):
```

Comment:
Is this actually deleting the instance? I think not. Rather it's clearing or resetting the data back to null. Another possibility is to have a delete flag attribute and mark that as True or as a timestamp

Reply:
Well, we're not exactly deleting the instance from the memory, but I thought the function could be used to showcase the use of private functions. I could delete the function or rename it to something else if you wish so 😄 .